src/shared/bootspec.c

```diff
@@ -755,45 +757,203 @@ static int boot_entry_load_unified(
 static int find_sections(
                 int fd,
                 const char *path,
-                char **ret_osrelease,
-                char **ret_cmdline) {
+                IMAGE_SECTION_HEADER **sections,
+                PeHeader **pe_header) {
 
-        _cleanup_free_ IMAGE_SECTION_HEADER *sections = NULL;
         _cleanup_free_ IMAGE_DOS_HEADER *dos_header = NULL;
-        _cleanup_free_ char *osrel = NULL, *cmdline = NULL;
-        _cleanup_free_ PeHeader *pe_header = NULL;
         int r;
 
         assert(fd >= 0);
         assert(path);
 
-        r = pe_load_headers(fd, &dos_header, &pe_header);
+        r = pe_load_headers(fd, &dos_header, pe_header);
         if (r < 0)
                 return log_warning_errno(r, "Failed to parse PE file '%s': %m", path);
 
-        r = pe_load_sections(fd, dos_header, pe_header, &sections);
+        r = pe_load_sections(fd, dos_header, *pe_header, sections);
         if (r < 0)
                 return log_warning_errno(r, "Failed to parse PE sections of '%s': %m", path);
 
+        return 0;
+}
+
+static int find_cmdline_section(
+                int fd,
+                const char *path,
+                IMAGE_SECTION_HEADER *sections,
+                PeHeader *pe_header,
+                char **ret_cmdline) {
+
+        int r;
+
+        if (!ret_cmdline)
+                return 0;
+
+        r = pe_read_section_data(fd, pe_header, sections, ".cmdline", PE_SECTION_SIZE_MAX, (void**) ret_cmdline, NULL);
+        if (r < 0 && r != -ENXIO) /* cmdline is optional */
+                return log_warning_errno(r, "Failed to read .cmdline section of '%s': %m", path);
+
+        return 0;
```

Comment:
as per coding style: our functions should strictly initialize all return params on success. returning 0 is success, hence you should initialize `ret_cmdline` here, i.e. to NULL.

Reply:
I don't get what you mean here. `ret_cmdline` is set by `pe_read_section_data`, why should it be initialized to NULL?

Reply:
If `pe_read_section_data` returns -ENXIO, it will not be initialized, no?